// @triframe/designer/src/paper/components/Menu/MenuItem.js
import color from 'color';
import * as React from 'react';
import { View, StyleSheet, } from 'react-native';
import Icon from '../Icon';
import TouchableRipple from '../TouchableRipple';
import Text from '../Typography/Text';
import { withTheme } from '../../core/theming';
import { black, white } from '../../styles/colors';
/**
 * A component to show a single list item inside a Menu.
 *
 */
let MenuItem = /** @class */ (() => {
    class MenuItem extends React.Component {
        render() {
            const { icon, title, disabled, onPress, theme, style, testID, titleStyle, } = this.props;
            const disabledColor = color(theme.dark ? white : black)
                .alpha(0.32)
                .rgb()
                .string();
            const titleColor = disabled
                ? disabledColor
                : color(theme.colors.text)
                    .alpha(0.87)
                    .rgb()
                    .string();
            const iconColor = disabled
                ? disabledColor
                : color(theme.colors.text)
                    .alpha(0.54)
                    .rgb()
                    .string();
            return (<TouchableRipple style={[styles.container, style]} onPress={onPress} disabled={disabled} testID={testID}>
        <View style={styles.row}>
          {icon ? (<View style={[styles.item, styles.icon]} pointerEvents="box-none">
              <Icon source={icon} size={24} color={iconColor}/>
            </View>) : null}
          <View style={[
                styles.item,
                styles.content,
                icon ? styles.widthWithIcon : null,
            ]} pointerEvents="none">
            <Text  style={[styles.title, { color: titleColor }, titleStyle]}>
              {title}
            </Text>
          </View>
        </View>
      </TouchableRipple>);
        }
    }
    MenuItem.displayName = 'Menu.Item';
    return MenuItem;
})();
const minWidth = 112;
const maxWidth = 280;
const iconWidth = 40;
const styles = StyleSheet.create({
    container: {
        paddingHorizontal: 8,
        minWidth,
        maxWidth,
        height: 48,
        justifyContent: 'center',
    },
    row: {
        flexDirection: 'row',
    },
    icon: {
        width: iconWidth,
    },
    title: {
        fontSize: 16,
    },
    item: {
        marginHorizontal: 8,
    },
    content: {
        justifyContent: 'center',
        minWidth: minWidth - 16,
        maxWidth: maxWidth - 16,
    },
    widthWithIcon: {
        maxWidth: maxWidth - (iconWidth + 48),
    },
});
export default withTheme(MenuItem);
// @component-docs ignore-next-line
export { MenuItem };
