// @triframe/mason/src/adapters/Heroku/index.js
import { Repo, print, exec, fs, path } from "../../utils"

export const Heroku = {

    requiresRelease: false,

    getOptions: async function() {

        await this.dependencyCheck()

        await this.findOrCreateRemote()

        await fs.writeFile(path.join(process.cwd(), 'Procfile'), `web: node ./Api.js`)

        return {}
    },

    dependencyCheck: async () => {
        try {
            await exec('heroku --version')
        } catch(err){
            throw Error(`heroku cli is required to use this utility. Install it here: https://devcenter.heroku.com/articles/heroku-cli#download-and-install, then login run 'heroku login'`)
        }
        try {
            await exec('heroku auth:token')
        } catch(err) {
            throw Error(`You are not logged into to heroku. Please run 'heroku login'`)
        }
    },

    deploy: async function(branch, options){
        let repo = new Repo

        print('Deploying...')
        let stuffForTesting = await repo.push(`heroku ${branch}:master`, {
            printOut: true
        })
        console.log(stuffForTesting)
    },

    async findOrCreateRemote(){
        let repo = new Repo

        let repos = await repo.getRemotes()

        if(!repos.includes('heroku')){
            await exec('heroku create')
            await exec('heroku addons:create heroku-postgresql:hobby-dev')
            await exec('heroku config:set NPM_CONFIG_PRODUCTION=true')
        }
    }

}
