// @triframe/designer/src/paper/components/Card/CardActions.js
import * as React from 'react';
import { StyleSheet, View } from 'react-native';
/**
 * A component to show a list of actions inside a Card.
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { Button, Card } from '../../';
 *
 * const MyComponent = () => (
 *   <Card>
 *     <Card.Actions>
 *       <Button>Cancel</Button>
 *       <Button>Ok</Button>
 *     </Card.Actions>
 *   </Card>
 * );
 *
 * export default MyComponent;
 * ```
 */
let CardActions = /** @class */ (() => {
    class CardActions extends React.Component {
        render() {
            return (<View {...this.props} style={[styles.container, this.props.style]}>
        {React.Children.map(this.props.children, child => React.isValidElement(child)
                ? React.cloneElement(child, {
                    compact: child.props.compact !== false,
                })
                : child)}
      </View>);
        }
    }
    CardActions.displayName = 'Card.Actions';
    return CardActions;
})();
const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'flex-start',
        padding: 8,
    },
});
export default CardActions;
