// @triframe/designer/src/DateInput/__web-components__/input-moment.js
import cx from 'classnames';
import React, { Component } from 'react';
import Calendar from './calendar';
import { TextInput } from '../../paper'
// import '../../../less/input-moment.css'
import { TouchableOpacity, Modal } from '../..';
import { Area } from '../../Layout';
import { Button } from '../../Button';
import { DateTime } from '@triframe/core'
import { Styles } from './Styles';

export default class InputMoment extends Component {
  static defaultProps = {
    value: DateTime.now(),
    label: 'Date Input',
    prevMonthIcon: 'chevron-left',
    nextMonthIcon: 'chevron-right',
    minStep: 1,
    hourStep: 1
  };

  state = {
    moment:  this.props.value === null ? DateTime.now() : this.props.value,
    cachedValue: this.props.value === null ? DateTime.now() : this.props.value,
    tab: 0,
    isActive: false
  };


  componentDidUpdate(){
    if(this.state.cachedValue != this.props.value){
      this.setState({
        cachedValue: this.props.value,
        moment: (this.props.value)
      })
    }
  }

  handleClickTab = (e, tab) => {
    e.preventDefault();
    this.setState({ tab: tab });
  };

  handleSave = e => {
    e.preventDefault();
    if (this.props.onSave) this.props.onSave();
  };

  render() {
    const { tab, isActive,  moment: m, } = this.state;
    let {
      className,
      prevMonthIcon,
      nextMonthIcon,
      minStep,
      hourStep,
      onChange,
      value,
      ...props
    } = this.props;
    const cls = cx('m-input-moment', className);
    if(value === null) value = DateTime.now()

    return (
      <>
      <Styles />
        <TouchableOpacity style={{ cursor: 'pointer' }} onPress={() => this.setState({ isActive: true })} >
          <TextInput value={(this.state.cachedValue).toFormat('MM/dd/yy')}  {...props} />
        </TouchableOpacity>
        <Modal visible={isActive} onDismiss={() => this.setState({ isActive: false })}>
          <Area alignX="center">
            <div className={cls}>
              <div className="tabs">
                <Calendar
                  moment={m}
                  onChange={moment => this.setState({ moment })}
                  prevMonthIcon={this.props.prevMonthIcon}
                  nextMonthIcon={this.props.nextMonthIcon}
                />
              </div>
              <Button icon="check" onPress={() => {
                this.setState({ isActive: false, moment: (this.state.cachedValue) })
                if(typeof onChange === 'function')  onChange(m)
              }}/>
            </div>
          </Area>
        </Modal>
      </>
    );
  }
}
