// @triframe/arbiter/src/serve/middleware.js
import expressCors from 'cors'
import bodyParser from 'body-parser'
import formidable from 'formidable'
import express from 'express'
import path from 'path'


export const corsMiddleware = expressCors({
    origin: function (origin, resolve) {
        if (origin && origin.includes('localhost')) {
            resolve(null, origin)
        } else {
            resolve(null, '')
        }
    },
    credentials: true
})

export const httpRedirectMiddleware = function(req, res, next){
    if (!req.hostname.includes('localhost') && req.headers['x-forwarded-proto'] === 'http') {
        res.redirect(`https://${req.hostname}${req.originalUrl}`)
    } else {
        next()
    }
}

export const formFileParsingMiddleware = function(req, res, next){
    const form = new formidable.IncomingForm();
    form.parse(req, (err, fields, files) => {
        req.files = Object.values(files)
        next()
    })
}

export const jsonParserMiddleware = bodyParser.json()

export const urlEncodedParserMiddleware = bodyParser.urlencoded({ extended: true })

export const staticFileMiddleware = express.static(path.resolve('./public'))
