// @triframe/arbiter/dist/serve/middleware.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.staticFileMiddleware = exports.urlEncodedParserMiddleware = exports.jsonParserMiddleware = exports.formFileParsingMiddleware = exports.httpRedirectMiddleware = exports.corsMiddleware = undefined;

var _cors = require("cors");

var _cors2 = _interopRequireDefault(_cors);

var _bodyParser = require("body-parser");

var _bodyParser2 = _interopRequireDefault(_bodyParser);

var _formidable = require("formidable");

var _formidable2 = _interopRequireDefault(_formidable);

var _express = require("express");

var _express2 = _interopRequireDefault(_express);

var _path = require("path");

var _path2 = _interopRequireDefault(_path);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const corsMiddleware = exports.corsMiddleware = (0, _cors2.default)({
  origin: function (origin, resolve) {
    if (origin && origin.includes('localhost')) {
      resolve(null, origin);
    } else {
      resolve(null, '');
    }
  },
  credentials: true
});

const httpRedirectMiddleware = exports.httpRedirectMiddleware = function (req, res, next) {
  if (!req.hostname.includes('localhost') && req.headers['x-forwarded-proto'] === 'http') {
    res.redirect(`https://${req.hostname}${req.originalUrl}`);
  } else {
    next();
  }
};

const formFileParsingMiddleware = exports.formFileParsingMiddleware = function (req, res, next) {
  const form = new _formidable2.default.IncomingForm();
  form.parse(req, (err, fields, files) => {
    req.files = Object.values(files);
    next();
  });
};

const jsonParserMiddleware = exports.jsonParserMiddleware = _bodyParser2.default.json();

const urlEncodedParserMiddleware = exports.urlEncodedParserMiddleware = _bodyParser2.default.urlencoded({
  extended: true
});

const staticFileMiddleware = exports.staticFileMiddleware = _express2.default.static(_path2.default.resolve('./public'));
