// @triframe/designer/src/paper/components/DataTable/DataTablePagination.js
import * as React from 'react';
import { StyleSheet, View, I18nManager, } from 'react-native';
import color from 'color';
import IconButton from '../IconButton';
import Text from '../Typography/Text';
import { withTheme } from '../../core/theming';
import Icon from '../Icon';
let DataTablePagination = /** @class */ (() => {
    class DataTablePagination extends React.Component {
        render() {
            const { label, page, numberOfPages, onPageChange, style, theme, ...rest } = this.props;
            const labelColor = color(theme.colors.text)
                .alpha(0.6)
                .rgb()
                .string();
            return (<View {...rest} style={[styles.container, style]}>
        <Text style={[styles.label, { color: labelColor }]} numberOfLines={1}>
          {label}
        </Text>
        <IconButton icon={({ size, color }) => (<Icon name="chevron-left" color={color} size={size} direction={I18nManager.isRTL ? 'rtl' : 'ltr'}/>)} color={theme.colors.text} disabled={page === 0} onPress={() => onPageChange(page - 1)}/>
        <IconButton icon={({ size, color }) => (<Icon name="chevron-right" color={color} size={size} direction={I18nManager.isRTL ? 'rtl' : 'ltr'}/>)} color={theme.colors.text} disabled={page === numberOfPages - 1} onPress={() => onPageChange(page + 1)}/>
      </View>);
        }
    }
    DataTablePagination.displayName = 'DataTable.Pagination';
    return DataTablePagination;
})();
const styles = StyleSheet.create({
    container: {
        justifyContent: 'flex-end',
        flexDirection: 'row',
        alignItems: 'center',
        paddingLeft: 16,
    },
    label: {
        fontSize: 12,
        marginRight: 44,
    },
});
export default withTheme(DataTablePagination);
// @component-docs ignore-next-line
export { DataTablePagination };
