// @triframe/designer/src/paper/components/DataTable/DataTableHeader.js
import * as React from 'react';
import color from 'color';
import { StyleSheet, View } from 'react-native';
import { black, white } from '../../styles/colors';
import { withTheme } from '../../core/theming';
let DataTableHeader = /** @class */ (() => {
    class DataTableHeader extends React.Component {
        render() {
            const { children, style, theme, ...rest } = this.props;
            const borderBottomColor = color(theme.dark ? white : black)
                .alpha(0.12)
                .rgb()
                .string();
            return (<View {...rest} style={[styles.header, { borderBottomColor }, style]}>
        {children}
      </View>);
        }
    }
    DataTableHeader.displayName = 'DataTable.Header';
    return DataTableHeader;
})();
const styles = StyleSheet.create({
    header: {
        flexDirection: 'row',
        height: 48,
        paddingHorizontal: 16,
        borderBottomWidth: StyleSheet.hairlineWidth * 2,
    },
});
export default withTheme(DataTableHeader);
// @component-docs ignore-next-line
export { DataTableHeader };
