// @triframe/mason/dist/adapters/Heroku/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Heroku = undefined;

var _utils = require("../../utils");

const Heroku = exports.Heroku = {
  requiresRelease: false,
  getOptions: async function () {
    await this.dependencyCheck();
    await this.findOrCreateRemote();
    await _utils.fs.writeFile(_utils.path.join(process.cwd(), 'Procfile'), `web: node ./Api.js`);
    return {};
  },
  dependencyCheck: async () => {
    try {
      await (0, _utils.exec)('heroku --version');
    } catch (err) {
      throw Error(`heroku cli is required to use this utility. Install it here: https://devcenter.heroku.com/articles/heroku-cli#download-and-install, then login run 'heroku login'`);
    }

    try {
      await (0, _utils.exec)('heroku auth:token');
    } catch (err) {
      throw Error(`You are not logged into to heroku. Please run 'heroku login'`);
    }
  },
  deploy: async function (branch, options) {
    let repo = new _utils.Repo();
    (0, _utils.print)('Deploying...');
    let stuffForTesting = await repo.push(`heroku ${branch}:master`, {
      printOut: true
    });
    console.log(stuffForTesting);
  },

  async findOrCreateRemote() {
    let repo = new _utils.Repo();
    let repos = await repo.getRemotes();

    if (!repos.includes('heroku')) {
      await (0, _utils.exec)('heroku create');
      await (0, _utils.exec)('heroku addons:create heroku-postgresql:hobby-dev');
      await (0, _utils.exec)('heroku config:set NPM_CONFIG_PRODUCTION=true');
    }
  }

};
