// @triframe/designer/src/paper/components/Appbar/AppbarBackAction.js
import * as React from 'react';
import { View, Image, I18nManager, StyleSheet, Platform, } from 'react-native';
import AppbarAction from './AppbarAction';
import Icon from '../Icon';
/**
 * A component used to display a back button in the appbar.
 */
let AppbarBackAction = /** @class */ (() => {
    class AppbarBackAction extends React.Component {
        render() {
            return (<AppbarAction {...this.props} icon={({ size, color }) => Platform.OS === 'ios' ? (<View style={[
                styles.wrapper,
                {
                    width: size,
                    height: size,
                    transform: [{ scaleX: I18nManager.isRTL ? -1 : 1 }],
                },
            ]}>
            </View>) : (<Icon name="arrow-left" color={color} size={size} direction={I18nManager.isRTL ? 'rtl' : 'ltr'}/>)}/>);
        }
    }
    AppbarBackAction.displayName = 'Appbar.BackAction';
    AppbarBackAction.defaultProps = {
        accessibilityLabel: 'Back',
    };
    return AppbarBackAction;
})();
const styles = StyleSheet.create({
    wrapper: {
        alignItems: 'center',
        justifyContent: 'center',
    },
    icon: {
        height: 21,
        width: 21,
        resizeMode: 'contain',
    },
});
export default AppbarBackAction;
